### PR TITLE
[HigherOrderOp] Better testing strategy for wrap that checks guards and recompiles

### LINKED ISSUE #109251: [HigherOrderOp] Better testing strategy for `wrap` that checks guards and recompiles
## Issue description

We've had some bugs in the past, like HOOs that don't guard on constant inputs. `wrap` is our guinea pig that we use to test the HOO mechanism. One suggestion is that we could beef up `_test_wrap_simple` to handle this:
- we could pass `_test_wrap_simple(f, args)` a function that generates the args instead of the args
- `_test_wrap_simple` compiles the wrap, and then runs it through with each of the generated args
- we check that each invocation is correct (compared to f(*args)).

### PR DESCRIPTION
**Pull request by yanboliang**
Fixes #109251

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: test/dynamo/test_higher_order_ops.py [hunk 1]
```diff
@@ -288,7 +304,7 @@ def f(x, y):
         y = 0.5
         actual_graph = self._test_wrap_simple(
             f,
-            (x, y),
+            default_args_generator((x, y)),
```
Comment (zou3519) [inline test/dynamo/test_higher_order_ops.py:307]: As written this just produces (x, y).  Ideally we would go through some of these (like this particular example) and change it to generate [(torch.randn(3, 1), 0.5), (torch.randn(3, 1), 0.4)] -- this way we'd test if we ended up specializing on a constant.

File: test/dynamo/test_higher_order_ops.py [hunk 2]
```diff
@@ -124,31 +128,40 @@ def _assert_wrap_fallback(self, func, args, setup=lambda: None):
...
+        # - other compilation metrics, e.g, # of ops in the dynamo captured graph,
+        #   the wrap has the expected number of args, etc
+        #
+        # we have one or multiple runs through with each of the args from args_generator,
+        # and we will check:
+        # - correctness and no graph breaks for every run
+        # - other compilation metrics only for the first run, since automatic_dynamic_shapes
+        #   may compile another dynamic version graph for the later runs
+        graph = None
+        for i, args in enumerate(args_generator):
```
Comment (zou3519) [inline test/dynamo/test_higher_order_ops.py:149]: Hmm, maybe `_test_wrap_simple` is doing too much. Do you have thoughts on alternate designs for the testing? The problem right now is that we wish to test a number of the wrap situations under different inputs to check that recompilation is correct. Maybe instead we want some OpInfo-like database that stores (HOO, sample inputs)

**Review by zou3519 — COMMENTED** [review]
This seems fine to me, I want to think about what would be a better testing design. There are a number of pain points I have with the current `wrap` testing:
- we don't test if we're guarding on the correct things. (This PR adds infrastructure to do this, but doesn't actually provide additional sample args in the generators to test this).
- It is really difficult to figure out what all of our wrap test cases are.

Maybe a better design could be we have a collection of wrap calls that are each associated with a sample inputs generator? Something like the following:

```
def wrap_capture_local(x, y):
    z = y

    def f(x):
        return x * y

    def func(x):
        return wrap(f, x)

    return func, (x,)

def wrap_capture_local_sample_inputs():
    return [(torch.randn(3), 4), (torch.randn(3), 5), (torch.randn(10), 4)]
```

And then we can generate a bunch of wrap tests using a parametrize decorator. There's an open question if this is in addition to all our current tests or if we try to hammer all of our existing tests into this design.

**Review by zou3519 — APPROVED** [review]

**Comment by ZainRizvi** [comment]
> @pytorchbot merge -f "irrelevant failures"

Fyi @yanboliang if you see irrelevant failures you can run "@pytorchbot merge -i". Mergebot will ignore any jobs that are currently failing on your PR while letting the rest of the PR checks and validations still run. It's a much safer alternative

**Pull request closed without merging**